File: ts/module-config/push-to-function/fs/boot/bar-bf-boot.ts
/*
 * Copyright © 2022 Paulo Villela. All rights reserved.
 * Use of this source code is governed by the MIT license
 * that can be found in the LICENSE file.
 */

import { AppCfgInfo} from "../../config/app-cfg";
import { barBfC, BarBfT, BarBfCfgInfo } from "../bar-bf";
import { mutableLifted } from "../../fwk/mutable-lifted";

function adapter(appCfg: AppCfgInfo): BarBfCfgInfo {
  return {
    z: appCfg.y
  };
}

export const barBfCfgAdapter = mutableLifted(adapter)

export function barBfBoot(appCfg: () => AppCfgInfo): BarBfT {
  return barBfC(barBfCfgAdapter(appCfg))
}

File: rs/pushtofunction_old/src/fs/boot/bar_bf_boot.rs
use crate::config::AppCfgInfo;
use crate::fs::bar_bf::{bar_bf_c, BarBfCfgInfo, BarBfCfgSrc, BarBfT};
use crate::fwk::const_or_adapt_by_ref;
use std::sync::{Arc, OnceLock};

fn bar_bf_cfg_adapter(app_cfg: &AppCfgInfo) -> BarBfCfgInfo {
    BarBfCfgInfo {
        u: app_cfg.y,
        v: app_cfg.x.clone(),
    }
}

pub static BAR_BF_CFG_INFO_OVERRIDE: OnceLock<BarBfCfgInfo> = OnceLock::new();

pub fn bar_bf_boot(app_cfg: fn() -> Arc<AppCfgInfo>) -> BarBfT {
    let get = const_or_adapt_by_ref(BAR_BF_CFG_INFO_OVERRIDE.get(), app_cfg, bar_bf_cfg_adapter);
    let bar_bf_cfg_src = BarBfCfgSrc { get };
    bar_bf_c(bar_bf_cfg_src)
}
